src/test/java/com/browserstack/examples/rule/WebDriverProviderRule.java

```diff
@@ -29,9 +40,44 @@ public String getMethodName() {
         return methodName;
     }
 
-    public WebDriver getWebDriver(WebDriverConfiguration webDriverConfiguration, Platform platform) {
+    public WebDriver getWebDriver(WebDriverConfiguration webDriverConfiguration, Platform platform) throws MalformedURLException {
         this.webDriverConfiguration = webDriverConfiguration;
-        // instantiate the WebDriver based on whether it is a remote or a local driver
+        switch (webDriverConfiguration.getDriverType()) {
+            case localDriver:
+                System.setProperty(WEBDRIVER_CHROME_DRIVER, Paths.get(platform.getDriverPath()).toString());
+                this.driver = new ChromeDriver();
+                break;
+            case remoteDriver:
+                RemoteDriverConfig remoteDriverConfig = webDriverConfiguration.getRemoteDriver();
+                CommonCapabilities commonCapabilities = remoteDriverConfig.getCommonCapabilities();
+                DesiredCapabilities platformCapabilities = new DesiredCapabilities();
+                if (StringUtils.isNoneEmpty(platform.getDevice())) {
+                    platformCapabilities.setCapability("device", platform.getDevice());
+                }
+                platformCapabilities.setCapability("browser", platform.getBrowser());
+                platformCapabilities.setCapability("browser_version", platform.getBrowserVersion());
+                platformCapabilities.setCapability("os", platform.getOs());
+                platformCapabilities.setCapability("os_version", platform.getOsVersion());
+                platformCapabilities.setCapability("name", getMethodName());
+                platformCapabilities.setCapability("project", commonCapabilities.getProject());
+                platformCapabilities.setCapability("build", commonCapabilities.getBuildPrefix());
+                commonCapabilities.getCapabilities().getCapabilityMap().forEach(platformCapabilities::setCapability);
+                if (platform.getCapabilities() != null) {
+                    platform.getCapabilities().getCapabilityMap().forEach(platformCapabilities::setCapability);
+                }
+                String user = remoteDriverConfig.getUser();
+                if (StringUtils.isNoneEmpty(System.getenv("BROWSERSTACK_USERNAME"))) {
+                    user = System.getenv("BROWSERSTACK_USERNAME");
+                }
+                String accessKey = remoteDriverConfig.getAccessKey();
+                if (StringUtils.isNoneEmpty(System.getenv("BROWSERSTACK_ACCESS_KEY"))) {
+                    accessKey = System.getenv("BROWSERSTACK_ACCESS_KEY");
+                }
+                platformCapabilities.setCapability("browserstack.user", user);
+                platformCapabilities.setCapability("browserstack.key", accessKey);
+                this.driver = new RemoteWebDriver(new URL(remoteDriverConfig.getHubUrl()), platformCapabilities);
+                break;
```

Comment:
This entire part can be moved to it's own method.

Reply:
done